neqo-transport/src/pace.rs

```diff
@@ -0,0 +1,136 @@
+    /// The maximum capacity, or burst size, in bytes.
+    m: usize,
+    /// The current capacity, in bytes.
+    c: usize,
+    /// The packet size or minimum capacity for sending, in bytes.
+    p: usize,
+}
+
+impl Pacer {
+    /// Create a new `Pacer`.  This takes the current time and the
+    /// initial congestion window.
+    ///
+    /// The value of `m` is the maximum capacity.  `m` primes the pacer
+    /// with credit and determines the burst size.  `m` must not exceed
+    /// the initial congestion window, but it should probably be lower.
+    ///
+    /// The value of `p` is the packet size, which determines the minimum
+    /// credit needed before a packet is sent.  This should be a substantial
+    /// fraction of the maximum packet size, if not the packet size.
+    pub fn new(now: Instant, m: usize, p: usize) -> Self {
+        assert!(m >= p, "maximum capacity has to be at least one packet");
+        Self { t: now, m, c: m, p }
+    }
+
+    /// Determine when the next packet will be available based on the provided RTT
+    /// and congestion window.  This doesn't update state.
+    /// This returns an option, with a value of `None` returned when there is
+    /// enough credit available to send.  That sounds counterintuitive, but it
+    /// avoids registering a zero-duration timer when there is credit for
+    /// multiple packets available, but fewer packets than that to send.
+    pub fn next(&self, rtt: Duration, cwnd: usize) -> Instant {
+        if self.c >= self.p {
+            qtrace!([self], "next {}/{:?} no wait: {:?}", cwnd, rtt, self.t);
+            self.t
+        } else {
+            // This is the inverse of the function in `spend`:
+            // self.t + rtt * (self.p - self.c) / (PACER_SPEEDUP * cwnd)
+            let r = rtt.as_nanos();
+            let d = r.saturating_mul(u128::try_from(self.p - self.c).unwrap());
+            let add = d / u128::try_from(cwnd * PACER_SPEEDUP).unwrap();
+            let dt = u64::try_from(add).map(Duration::from_nanos).unwrap_or(rtt);
+            qtrace!(
+                [self],
+                "next {}/{:?} wait {:?}: {:?}",
+                cwnd,
+                rtt,
+                dt,
+                self.t + dt
+            );
+            self.t + dt
+        }
+    }
+
+    /// Spend credit.  This cannot fail; users of this API are expected to call
+    /// next() to determine when to spend.  This takes the current time (`now`),
+    /// an estimate of the round trip time (`rtt`), the estimated congestion
+    /// window (`cwnd`), and the number of bytes that were sent (`count`).
+    pub fn spend(&mut self, now: Instant, rtt: Duration, cwnd: usize, count: usize) {
+        qtrace!([self], "spend {} over {}, {:?}", count, cwnd, rtt);
+        // Increase the capacity by:
+        //    `(now - self.t) * PACER_SPEEDUP * cwnd / rtt`
+        // That is, the elapsed fraction of the RTT times rate that data is added.
+        let incr = (now - self.t)
```

Comment:
```suggestion
        let incr = now.saturating_duration_since(self.t)
```

Otherwise, if `now` is in the past relative to `self.t`, bad things happen.

Getting the overflows and underflows right on this is nasty.